_posts/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.md

```diff
@@ -0,0 +1,110 @@
+---
+layout: post
+asset-type: post
+name: 99% code coverage - Do we have a good safety net to change this legacy code?
+title: '99% code coverage - Do we have a good safety net to change this legacy code?'
+date: 2017-08-29 12:00:00 +00:00
+author: Raquel M Carmena
+image:
+  src: /assets/img/custom/blog/2017-08-29-do-we-have-a-good-safety-net-to-change-this-legacy-code.jpg
+tags:
+- coverage
+- testing
+- mutation testing
+- quality
+
+---
+A long time ago, I met a development team which was working under big pressure by the quality team. Personally, I don’t like this kind of differences between development and quality teams, because it leads to development teams not feeling responsible for quality and to a confrontational relationship. They should be working collaboratively towards a unified goal of delivering a quality product.
+
+One of the requirements was to have more than 85% of code coverage to ensure code quality. The result was perverse: development team wrote tests without assertions; they only invoked methods with different arguments to reach the desired percentage. It’s clear that they didn’t follow TDD.
+
+Code coverage only gives us information about the percentage of code lines which are executed during tests. Let’s see a way to check if we have a good safety net with current tests to make changes in production code.
+
+If we made a change in production code, such as “replacing + with -”, “replacing < with >=” or “returning a different value”, test battery might detect that change. In order words, tests might fail.
+
+There are tools to make changes in production code automatically and to run tests in order to check if those changes are detected. It’s talked about:
+
+* **Mutators**: changes to be applied
+* **Mutations**: new versions of production code after applying mutators
+* **Mutations are killed**: tests fail; mutators are detected
+* **Mutations survive**: tests don’t fail; mutators aren't detected
+
+So, we should aim at killing every mutation with tests. 
+
+When I heard about it I thought about that game I played when I was just a teenager: _Super Pang_.
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/super-pang-game.jpg" alt="Super Pang Game" class="img img-responsive"/>
+</center>
+<br/>
+
+And I imagined a situation such as this:
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-08-29-coverage/super-pang-mutation-testing.png" alt="Super Pang - Mutation Testing" class="img img-responsive"/>
+</center>
+
+It’s called **mutation testing** and it's a good way to make sure that you have a good safety net with your current tests to refactor production code or to add new features, among others.
```

Comment:
`among others.`?
You mean among other things?
You mean adding a feature to existing set of features?

Reply:
`among other things`, but I can remove it. It's not necessary. Thank you @halimakoundi !